src/gaxios.ts

```diff
@@ -53,9 +51,62 @@ function loadProxy() {
   if (proxy) {
     HttpsProxyAgent = require('https-proxy-agent');
   }
-  return proxy;
 }
-loadProxy();
+
+function matchingProxyStrings(
+  envVarHTTPS: string | undefined,
+  envVarHTTP: string | undefined,
+  envVarhttps: string | undefined,
+  envVarhttp: string | undefined,
+  url: string
+) {
+  const arrayOfEnvVariables = (
+    envVarHTTPS ||
+    envVarHTTP ||
+    envVarhttps ||
+    envVarhttp
+  )?.split(',');
+
+  let isMatch;
+  if (arrayOfEnvVariables && arrayOfEnvVariables.length > 0) {
+    const parsedURL = new URL(url);
+    isMatch = arrayOfEnvVariables.find(url => {
+      if (url.startsWith('*.') || url.startsWith('.')) {
+        url = url.replace('*', '');
+        return parsedURL.hostname.endsWith(url);
+      } else {
+        return url === parsedURL.origin || url === parsedURL.hostname;
+      }
+    });
+  }
+  return isMatch;
+}
+
+// Figure out if we should be using a proxy. Only if it's required, load
+// the https-proxy-agent module as it adds startup cost.
+function getProxy(url: string) {
+  loadProxy();
+  const shouldThisBeNoProxy = matchingProxyStrings(
+    process.env.no_proxy,
+    process.env.no_proxy,
+    undefined,
+    undefined,
+    url
+  );
+  // If there is a match between the no_proxy env variables and the url, then do not proxy
+  if (shouldThisBeNoProxy) {
+    return undefined;
+    // If there is not a match between the no_proxy env variables and the url, check to see if there should be a proxy
+  } else {
+    return matchingProxyStrings(
+      process.env.HTTPS_PROXY,
+      process.env.https_proxy,
+      process.env.HTTP_PROXY,
+      process.env.http_proxy,
+      url
+    );
```

Comment:
we should just return the proxy here, the logic is different for `http_proxy`, vs., `no_proxy` (one contains a list of URLs, the other is just a single URL representing the proxy).